include/wsrep/buffer.hpp

```diff
@@ -38,35 +48,59 @@ namespace wsrep
             , size_(size)
         { }
 
+        const_buffer(const const_buffer& b)
+            : ptr_(b.ptr())
+            , size_(b.size())
+        { }
+
         const void* ptr() const { return ptr_; }
-        const void* data() const { return ptr_; }
+        const char* data() const { return static_cast<const char*>(ptr_); }
         size_t size() const { return size_; }
 
+        const_buffer& operator=(const const_buffer& b)
+        {
+            ptr_  = b.ptr();
+            size_ = b.size();
+            return *this;
+        }
     private:
-        // const_buffer(const const_buffer&);
-        // const_buffer& operator=(const const_buffer&);
         const void* ptr_;
         size_t size_;
     };
 
 
-    class mutable_buffer
+    class mutable_buffer : public buffer
     {
     public:
         mutable_buffer()
             : buffer_()
         { }
 
+        void resize(size_t s) { buffer_.resize(s); }
+
+        void push_back(const std::string& str)
```

Comment:
Not if `str` is a temporary. At least for me it results in a "pure virtual method called" error.
So, I'd leave it for convenience.

Reply:
I would be very surprised if the above would not work. Maybe you were doing something like
```
std::ostringstream os;
os << "stuff";
buf.push_back(os.str().data(), os.str().data() + os.str().size());
```
where `os.str()` returns by *value* and the `data()` method is called for two different objects?

In any case, I don't believe this is the right place for convenience method. What if the user wants to use `std::vector<char>` instead of `std::string`, or even some custom container? We should again expand the class for convenience.

The `push_back(const char*, const char*)` is generic enough to work in most of the cases. The convenience methods should be implemented as free functions, and preferably in compilation unit where they are used. Pushing convenience methods in public headers is IMO wrong approach, since it bloats the public API and usually cannot be generic enough to fulfill all needs.

Reply:
Or, maybe the overload could be made template method:
´``
template <class C> void push_back(const C& c)
{
    std::copy(c.begin(), c.end(), std::back_inserter(buffer_));
}
```

In this case it could be generic enough to be justified (should work for `std::vector<char>`, `std::string`, `std::array<char>`.

Reply:
awesome!